File: 3. Behavioral Design Patterns/Chain of Responsibility/Python/ChainOfResponsibilityTemplate.py
# Chain of Responsibility Pattern – Python Template

from abc import ABC, abstractmethod

class Handler(ABC):
    def __init__(self):
        self.next_handler = None
    
    def set_next(self, handler):
        self.next_handler = handler
    
    @abstractmethod
    def handle_request(self, request):
        pass

class Request:
    def __init__(self, data):
        self.data = data

class ConcreteHandlerA(Handler):
    def handle_request(self, request):
        if self.can_handle(request):
            return self.process_request(request)
        elif self.next_handler:
            return self.next_handler.handle_request(request)
        return "No handler found"
    
    def can_handle(self, request):
        return True  # condition
    
    def process_request(self, request):
        return f"ConcreteHandlerA processed: {request.data}"

class ConcreteHandlerB(Handler):
    def handle_request(self, request):
        if self.can_handle(request):
            return self.process_request(request)
        elif self.next_handler:
            return self.next_handler.handle_request(request)
        return "No handler found"
    
    def can_handle(self, request):
        return True  # condition
    
    def process_request(self, request):
        return f"ConcreteHandlerB processed: {request.data}"

def main():
    handler_a = ConcreteHandlerA()
    handler_b = ConcreteHandlerB()
    handler_a.set_next(handler_b)
    
    request = Request("test_data")
    print(handler_a.handle_request(request))

if __name__ == "__main__":
    main()

File: 3. Behavioral Design Patterns/Chain of Responsibility/Go/ChainOfResponsibilityTemplate.go
package main

// Chain of Responsibility Pattern – Go Template

type Handler interface {
	SetNext(Handler)
	HandleRequest(*Request)
}

type Request struct{ Data string }

type BaseHandler struct{ next Handler }
func (h *BaseHandler) SetNext(handler Handler) { h.next = handler }

type ConcreteHandlerA struct{ BaseHandler }
func (h *ConcreteHandlerA) HandleRequest(req *Request) {
	if h.canHandle(req) {
		h.processRequest(req)
	} else if h.next != nil {
		h.next.HandleRequest(req)
	}
}
func (h *ConcreteHandlerA) canHandle(req *Request) bool { return true /* condition */ }
func (h *ConcreteHandlerA) processRequest(req *Request) { /* handle */ }

type ConcreteHandlerB struct{ BaseHandler }
func (h *ConcreteHandlerB) HandleRequest(req *Request) {
	if h.canHandle(req) {
		h.processRequest(req)
	} else if h.next != nil {
		h.next.HandleRequest(req)
	}
}
func (h *ConcreteHandlerB) canHandle(req *Request) bool { return true /* condition */ }
func (h *ConcreteHandlerB) processRequest(req *Request) { /* handle */ }

func main() {
	handlerA := &ConcreteHandlerA{}
	handlerB := &ConcreteHandlerB{}
	handlerA.SetNext(handlerB)
	handlerA.HandleRequest(&Request{})
}
